network/kademlia_load_balancer.go

```diff
@@ -0,0 +1,256 @@
+		klb.initCountFunc = klb.nearestNeighbourUseCount
+	} else {
+		klb.initCountFunc = klb.leastUsedCountInBin
+	}
+
+	go klb.listenNewPeers()
+	go klb.listenOffPeers()
+	return klb
+}
+
+// Consumer functions. A consumer is a function that uses an element returned by an iterator. It usually also returns
+// a boolean signaling if it wants to iterate more or not. We created an alias for consumer function (LBBinConsumer)
+// for code clarity.
+
+// An LBPeer represents a peer with a AddUseCount() function to signal that the peer has been used in order
+// to account it for LB sorting criteria.
+type LBPeer struct {
+	Peer  *Peer
+	stats *resourceusestats.ResourceUseStats
+}
+
+// AddUseCount is called to account a use for these peer. Should be called if the peer is actually used.
+func (lbPeer LBPeer) AddUseCount() {
+	lbPeer.stats.AddUse(lbPeer.Peer)
+}
+
+// LBBin represents a Bin of LBPeer's
+type LBBin struct {
+	LBPeers        []LBPeer
+	ProximityOrder int
+}
+
+// LBBinConsumer will be provided with a list of LBPeer's in LB criteria ordering (currently in least used ordering).
+// Should return true if it must continue iterating LBBin's or stops if false.
+type LBBinConsumer func(bin LBBin) bool
+
+// KademliaLoadBalancer tries to balance request to the peers in Kademlia returning the peers sorted
+// by least recent used whenever several will be returned with the same po to a particular address.
+// The user of KademliaLoadBalancer should signal if the returned element (LBPeer) has been used with the
+// function lbPeer.AddUseCount()
+type KademliaLoadBalancer struct {
+	kademlia         KademliaBackend                    // kademlia to obtain bins of peers
+	resourceUseStats *resourceusestats.ResourceUseStats // a resourceUseStats to count uses
+	onPeerSub        *pubsubchannel.Subscription        // a pubsub channel to be notified of new peers in kademlia
+	offPeerSub       *pubsubchannel.Subscription        // a pubsub channel to be notified of removed peers in kademlia
+	quitC            chan struct{}
+
+	initCountFunc func(peer *Peer, po int) int //Function to use for initializing a new peer count
+}
+
+// Stop unsubscribe from notifiers
+func (klb *KademliaLoadBalancer) Stop() {
+	klb.onPeerSub.Unsubscribe()
+	klb.offPeerSub.Unsubscribe()
+	close(klb.quitC)
+}
+
+// EachBinNodeAddress calls EachBinDesc with the base address of kademlia (the node address)
+func (klb *KademliaLoadBalancer) EachBinNodeAddress(consumeBin LBBinConsumer) {
+	klb.EachBinDesc(klb.kademlia.BaseAddr(), consumeBin)
+}
+
+// EachBinFiltered returns all bins in descending order from the perspective of base address.
+// Only peers with the provided capabilities capKey are considered.
+// All peers in that bin will be provided to the LBBinConsumer sorted by least used first.
+func (klb *KademliaLoadBalancer) EachBinFiltered(base []byte, capKey string, consumeBin LBBinConsumer) {
+	_ = klb.kademlia.EachBinDescFiltered(base, capKey, 0, func(peerBin *PeerBin) bool {
```

Comment:
Error is not handled. Even it can be only one error that is ignored now, if the function returns an error, it should be handled, at least logged.